File: panthera/addon/synthDrivers/_panthera/updates.py
# -*- coding: utf-8 -*-
"""Is there a newer release than the one running?

Asked for by Sean, who wanted a check-for-updates button rather than having
to watch the repository.

**Only ever when somebody presses the button.**  Nothing here runs on a timer
and nothing runs at start-up.  A screen reader that quietly contacts a server
every launch is telling that server when its owner sits down at their machine,
and nobody asked it to.  The button is the consent.

**And never on a secure screen.**  There NVDA is SYSTEM on the sign-in
desktop; reaching onto the network as SYSTEM, from a dialog nobody can be
logged in to have opened, is not a thing this add-on should be able to do.
The caller hides the button there; this module refuses as well, because a
guard that lives only in the caller is a guard that moves.

The version comparison is pure and the fetch is not, so the comparison is
tested and the fetch is kept to the smallest piece of code that can be wrong.
"""
import os
import re

#: GitHub's own "newest published release" endpoint.  It excludes drafts and
#: prereleases without being asked to, which is exactly the wanted behaviour:
#: a draft under test must never advertise itself to everybody.
LATEST_API = ("https://api.github.com/repos/tgeczy/panthera-speech"
              "/releases/latest")

#: Where to send somebody who wants it.  The human page, not the API.
LATEST_PAGE = "https://github.com/tgeczy/panthera-speech/releases/latest"

#: Tags are `pantheraspeech/v2.0.0`; older ones and hand-typed ones may be
#: `v2.0.0` or `2.0.0`.  Anything after the numbers -- `-rc1`, `-beta` -- is
#: matched but not compared, because a prerelease is not offered by this
#: endpoint anyway and a suffix should never make a version look *newer*.
_VERSION = re.compile(r"(\d+(?:\.\d+)*)")


def parse_version(text):
    """-> a tuple of ints, or None if there is no version in `text`.

    Trailing zeroes are not trimmed: (2, 0) and (2, 0, 0) compare equal
    anyway once padded, and padding is done in `is_newer` where both sides
    are known.
    """
    if not text:
        return None
    found = _VERSION.search(str(text))
    if not found:
        return None
    try:
        return tuple(int(part) for part in found.group(1).split("."))
    except ValueError:
        return None


def is_newer(latest, installed):
    """-> True when `latest` is a strictly higher version than `installed`.

    Padded to the same length so 2.1 beats 2.0.9 and 2.0 ties 2.0.0.  Either
    side unparseable is False: an add-on that cannot tell should say nothing
    rather than announce an update that may not exist.
    """
    a, b = parse_version(latest), parse_version(installed)
    if not a or not b:
        return False
    width = max(len(a), len(b))
    a = a + (0,) * (width - len(a))
    b = b + (0,) * (width - len(b))
    return a > b


def installed_version():
    """-> the running add-on's version, or None.

    `addonHandler` is the right answer inside NVDA and absent everywhere else,
    so the manifest beside this file is the fallback -- and the manifest is
    what `addonHandler` would have read anyway.
    """
    try:
        import addonHandler
        addon = addonHandler.getCodeAddon()
        if addon and addon.version:
            return addon.version
    except Exception:
        pass
    here = os.path.dirname(os.path.abspath(__file__))
    manifest = os.path.join(os.path.dirname(os.path.dirname(here)),
                            "manifest.ini")
    try:
        with open(manifest, encoding="utf-8") as f:
            for line in f:
                if line.strip().startswith("version"):
                    return line.split("=", 1)[1].strip()
    except (OSError, IndexError):
        pass
    return None


def _secure():
    """-> True on a secure screen, and True if we cannot tell."""
    try:
        import globalVars
        return bool(globalVars.appArgs.secure)
    except Exception:
        # Outside NVDA entirely -- the tests, a command line -- which is not a
        # secure screen.  Only an *NVDA* that will not answer is treated as
        # one, and there is no such thing.
        return False


def latest_release(timeout=10, opener=None):
    """-> (versionString, pageUrl, addonUrl) -- or (None, reason, None).

    `addonUrl` is the `.nvda-addon` asset itself, when the release carries
    exactly the file NVDA installs; the caller downloads it and hands it to
    NVDA, whose own install dialog is still the thing that asks.  Sending
    somebody to a web page to find the right link themselves is homework a
    screen reader user did not ask for -- Tomi's words were less polite --
    so the page URL is only the fallback for a release with no such asset.

    `opener` exists for the tests: anything callable that takes a URL and
    returns bytes.  The default reaches the network, which is why it is the
    only part of this module that is not tested.
    """
    if _secure():
        return None, "not while NVDA is on a secure screen", None
    try:
        import json
        if opener is None:
            import urllib.request

            def opener(url):
                request = urllib.request.Request(url, headers={
                    # GitHub refuses an unidentified caller, and this says
                    # who we are without saying anything about the machine.
                    "User-Agent": "panthera-speech-addon",
                    "Accept": "application/vnd.github+json",
                })
                with urllib.request.urlopen(request, timeout=timeout) as r:
                    return r.read()

        payload = json.loads(opener(LATEST_API).decode("utf-8"))
    except Exception as e:
        return None, (str(e) or e.__class__.__name__), None
    tag = payload.get("tag_name") or payload.get("name")
    if not parse_version(tag):
        return None, "the newest release does not name a version", None
    addon = None
    for asset in payload.get("assets") or []:
        name = asset.get("name") or ""
        if name.endswith(".nvda-addon") and asset.get("browser_download_url"):
            addon = asset["browser_download_url"]
            break
    return tag, (payload.get("html_url") or LATEST_PAGE), addon


def fetch(url, timeout=60):
    """Download `url` into %TEMP% and -> the file's path.

    The old file of the same name goes first, so a half-written download
    from a failed attempt is never the thing NVDA is handed.  Exceptions
    propagate: the caller is on a worker thread and turns them into words.
    """
    import tempfile
    import urllib.request

    name = url.rsplit("/", 1)[-1]
    if not name.endswith(".nvda-addon"):
        name = "panthera-update.nvda-addon"
    path = os.path.join(tempfile.gettempdir(), name)
    try:
        if os.path.isfile(path):
            os.remove(path)
    except OSError:
        pass
    request = urllib.request.Request(url, headers={
        "User-Agent": "panthera-speech-addon"})
    with urllib.request.urlopen(request, timeout=timeout) as r:
        data = r.read()
    with open(path, "wb") as f:
        f.write(data)
    return path

File: src/platforms/android/app/src/main/kotlin/com/pantheraspeech/tts/Updates.kt
package com.pantheraspeech.tts

import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

/**
 * Is there a newer APK than the one running?
 *
 * Asked for in panthera-speech#19. The answer stops at pointing, on purpose.
 * An app cannot install an APK unless the person has switched on "install
 * unknown apps" for it, one app at a time, and asking for that is the sort
 * of thing Google treats third-party apps badly for. So the check finds the
 * release and hands its APK to the browser: the browser downloads it,
 * opening the download installs it, and Android asks once to let the
 * *browser* install apps. Slower than an updater, and it asks for nothing.
 *
 * **Only ever when somebody presses the button.** The desktop add-on's rule,
 * kept for the same reason: an engine that quietly contacts a server on its
 * own tells that server when its owner picked the phone up.
 *
 * `/releases/latest` is not consulted. That is pinned to the newest release
 * carrying the desktop files, so the NVDA and SAPI updaters stay quiet about
 * Android-only releases; this walks the recent releases for the newest one
 * that carries an APK. The version comparison is pure and tested; the fetch
 * is the smallest piece that reaches the network.
 */
object Updates {
    const val RELEASES_API =
        "https://api.github.com/repos/tgeczy/panthera-speech/releases?per_page=20"
    const val RELEASES_PAGE = "https://github.com/tgeczy/panthera-speech/releases"

    class Release(val version: String, val page: String, val apk: String) {
        /** "3.1.0" out of "pantheraspeech/v3.1.0". */
        val number: String get() = parseVersion(version)?.joinToString(".") ?: version
    }

    sealed class Answer {
        class UpToDate(val installed: String) : Answer()
        class Available(val release: Release, val installed: String) : Answer()
        class Failed(val reason: String) : Answer()
    }

    private val VERSION = Regex("""(\d+(?:\.\d+)*)""")

    /** -> the numbers in a version, or null if there are none. */
    fun parseVersion(text: String?): List<Int>? {
        val found = text?.let { VERSION.find(it) } ?: return null
        return try { found.groupValues[1].split('.').map { it.toInt() } }
            catch (e: NumberFormatException) { null }
    }

    /** True when `latest` is strictly higher than `installed`. Padded, so
     * 3.1 beats 3.0.9 and 3.0 ties 3.0.0. Either side unreadable is false:
     * a check that cannot tell should say nothing rather than announce an
     * update that may not exist. */
    fun isNewer(latest: String?, installed: String?): Boolean {
        val a = parseVersion(latest) ?: return false
        val b = parseVersion(installed) ?: return false
        val width = maxOf(a.size, b.size)
        val pa = a + List(width - a.size) { 0 }
        val pb = b + List(width - b.size) { 0 }
        for (i in 0 until width) if (pa[i] != pb[i]) return pa[i] > pb[i]
        return false
    }

    /** The newest published release in the API's answer that carries an APK,
     * or null when none does. Drafts and pre-releases are passed over. */
    fun newestApk(json: String): Release? {
        val releases = org.json.JSONArray(json)
        var best: Release? = null
        for (i in 0 until releases.length()) {
            val r = releases.getJSONObject(i)
            if (r.optBoolean("draft", false) || r.optBoolean("prerelease", false)) continue
            val tag = if (r.has("tag_name")) r.getString("tag_name") else r.optString("name", "")
            if (parseVersion(tag) == null) continue
            val assets = r.optJSONArray("assets") ?: continue
            var apk: String? = null
            for (j in 0 until assets.length()) {
                val a = assets.getJSONObject(j)
                if (a.optString("name", "").endsWith(".apk", ignoreCase = true)
                        && a.has("browser_download_url")) {
                    apk = a.getString("browser_download_url")
                    break
                }
            }
            if (apk == null) continue
            val page = if (r.has("html_url")) r.getString("html_url") else RELEASES_PAGE
            if (best == null || isNewer(tag, best.version)) best = Release(tag, page, apk)
        }
        return best
    }

    /** The one piece that reaches the network. */
    fun fetch(url: String, timeoutMs: Int = 10000): String {
        val c = URL(url).openConnection() as HttpURLConnection
        try {
            c.connectTimeout = timeoutMs
            c.readTimeout = timeoutMs
            c.setRequestProperty("User-Agent", "panthera-speech-android")
            c.setRequestProperty("Accept", "application/vnd.github+json")
            if (c.responseCode !in 200..299) throw IOException("HTTP ${c.responseCode}")
            return c.inputStream.bufferedReader().use { it.readText() }
        } finally {
            c.disconnect()
        }
    }

    /** `opener` exists for the tests: anything that takes a URL and returns
     * the body. The default reaches the network. */
    fun check(installed: String, opener: (String) -> String = { fetch(it) }): Answer {
        val json = try { opener(RELEASES_API) } catch (e: Exception) {
            return Answer.Failed(e.message?.takeIf { it.isNotBlank() } ?: e.javaClass.simpleName)
        }
        val newest = try { newestApk(json) } catch (e: Exception) {
            return Answer.Failed("the release list could not be read")
        } ?: return Answer.Failed("no release carries an APK")
        return if (isNewer(newest.version, installed)) Answer.Available(newest, installed)
               else Answer.UpToDate(installed)
    }
}
